src/cls/IPM/Utils/Migration.cls

```diff
@@ -0,0 +1,108 @@
+Class %IPM.Utils.Migration
+{
+
+ClassMethod RunAll(verbose As %Boolean = 1) As %Status
+{
+    Set sc = $$$OK
+    Try {
+        Do ..MigrateZPMToIPM(verbose)
+    } Catch e {
+        Set sc = e.AsStatus()
+    }
+    Quit sc
+}
+
+ClassMethod MigrateZPMToIPM(verbose As %Boolean = 1)
+{
+    If '$$$comClassDefined("%ZPM.PackageManager.Developer.Module") {
+        Write:verbose !,"Older IPM version not found; nothing to migrate.",!
+        Quit
+    }
+
+    Set sc = $$$OK
+    Set initTLevel = $TLevel
+    // Intentionally runs for current namespace only
+    Try {
+        TSTART
+
+        // ^ZPM.Dev.ModuleD -> ^IPM.Storage.ModuleD
+        Set oldId = ""
+        For {
+            Set oldId = $order(^ZPM.Dev.ModuleD(oldId),1,moduleData)
+            Quit:oldId=""
+
+            Set name = $ListGet(moduleData,2)
+            If ##class(%IPM.Storage.Module).NameExists(name) {
+                Write:verbose !,"Package already found in new storage: ",name
+                Continue
+            }
+
+            Write:verbose !,"Migrating storage for ",name," ... "
+
+            // Bring everything over - storage definitions are fortunately compatible.
+            Kill moduleData
+            Merge moduleData = ^ZPM.Dev.ModuleD(oldId)
+
+            // In ^IPM.Storage.ModuleD(<ID>,"Defaults",n) = $lb($lb(<data>),classname), map classname
+            // to equivalent - %ZPM.PackageManager.Developer.ModuleSetting -> %IPM.Storage.ModuleSetting
+            Set defaultKey = ""
+            For {
+                Set defaultKey = $Order(moduleData("Defaults",defaultKey),1,defaultData)
+                Quit:defaultKey=""
+                Set moduleData("Defaults",defaultKey) = $ListBuild($ListGet(defaultData,1),$Replace($ListGet(defaultData,2),"%ZPM.PackageManager.Developer.ModuleSetting","%IPM.Storage.ModuleSetting"))
+            }
+
+            Set newId = $Increment(^IPM.Storage.ModuleD)
+            Merge ^IPM.Storage.ModuleD(newId) = moduleData
+
+            // Make sure loading the object works
+            Set newObj = ##class(%IPM.Storage.Module).%OpenId(newId,,.sc)
+            $$$ThrowOnError(sc)
+
+            // Save object to validate
+            $$$ThrowOnError(newObj.%Save())
+
+            Write:verbose "done."
+        }
+
+        $$$ThrowOnError(##class(%IPM.Storage.Module).%BuildIndices(,1,1))
+
+        // Migrate %ZPM.PackageManager.Developer.Document.ModuleStream to %IPM.StudioDocument.ModuleStream
+        // ^ZPM.Dev.ModuleDocD -> ^IPM.StudioDoc.ModuleStreamD
+        // Uses objects for the actual data migration to more cleanly handle the stream.
+        Set oldId = ""
+        For {
+            Set oldId = $order(^ZPM.Dev.ModuleDocD(oldId),1,moduleDocData)
```

Comment:
Rather than separate for loops, as IDs are identified in the module storage, also migrate its corresponding studio document?